File: tp-8/Nave/haskell/Sector.hs
module Sector (
    Sector, 
    crearS,
    sectorId,
    componenteS,
    tripulantesS,
    agregarC,
    agregarT
) where

import Set
import Tipos


data Sector = S SectorId (Set Nombre) [Componente]  deriving Show


{-
    Proposito: 
    PRECONDICION:   
    COSTO:  O(1) constante
-}
crearS :: SectorId -> Sector
crearS sId = (S sId emptyS [])

{-
    Proposito: 
    PRECONDICION:   
    COSTO: O(1) constante
-}
sectorId :: Sector -> SectorId 
sectorId (S sId _ _) = sId

{-
    Proposito: 
    PRECONDICION:   
    COSTO: O(1) constante
-}
componenteS :: Sector -> [Componente]
componenteS (S sId _ cs) = cs


{-
    Proposito: 
    PRECONDICION:   
    COSTO: O(1) constante
-}
tripulantesS :: Sector -> Set Nombre
tripulantesS (S sId ts _) = ts

{-
    Proposito: 
    PRECONDICION:   
    COSTO:  O(1) constante
-}
agregarC :: Componente -> Sector -> Sector
agregarC c (S sId ts cs) = (S sId ts (c:cs))

{-
    Proposito: 
    PRECONDICION:   
    COSTO:  O(log T) siendo T la longitud de tripulantes
-}
agregarT :: Nombre -> Sector -> Sector
agregarT n (S sId ts cs) =  (S sId (addS n ts) cs)

File: tp-8/Nave/cplusplus/sector/Sector.cpp
#include <iostream>
#include "Sector.h"
using namespace std;


//costo: constante O(1)
Sector crearS(SectorId sId){
    Sector s = new SectorSt;
    s->sectorId = sId;
    s->componentes = vector<Componente>();
    s->tripulantes = emptyS();
    return s;
};

//costo: constante O(1)
SectorId sectorId(Sector s){
    return s->sectorId;
};

//costo: constante O(1)
Componente[] componentesS (Sector s){
    return s->componentes;
};

//costo: constante O(1)
Set tripulantesS (Sector s){
    return s->tripulantes;
};

//costo: O(n) donde n es la cantidad de componentes del sector
void agregarC (Componente c, Sector s){
    Set componentes = s->componentes;
    componentes = addS(c, componentes);
    s->componentes = componentes;
};

//costo: O(n) donde n es la cantidad de tripulantes del sector
void agregarT (Tripulant t, Sector s){
    Set tripulantes = s->tripulantes;
    tripulantes = addS(t, tripulantes);
    s->tripulantes = tripulantes;
};
